oikeustulkkirekisteri-service/src/main/java/fi/vm/sade/oikeustulkkirekisteri/service/impl/EmailNotificationServiceImpl.java

```diff
@@ -89,84 +90,150 @@
+            throw e;
+        }
+    }
+
+    private SahkopostiMuistutus createSahkopostiMuistutus(Oikeustulkki oikeustulkki, String vastaanottaja) {
         SahkopostiMuistutus muistutus = new SahkopostiMuistutus();
         muistutus.setOikeustulkki(oikeustulkki);
         muistutus.setLahettaja(senderEmail);
+        muistutus.setVastaanottaja(vastaanottaja);
         muistutus.setTemplateName(templateName);
-        oikeustulkki.getSahkopostiMuistutukset().add(muistutus);
-        HenkiloRestDto henkilo = found(oikeustulkkiCacheService.findHenkiloByOid(oikeustulkki.getTulkki().getHenkiloOid()));
-        muistutus.setVastaanottaja(findOikeustulkkiYhteystietoArvo(henkilo, YHTEYSTIETO_SAHKOPOSTI)
-                .orElseThrow(() -> new IllegalStateException("Oikeustulkki henkiloOid="+oikeustulkki.getTulkki().getHenkiloOid()
-                        + " does not have work email.")));
         muistutus.setKieli(new Kieli(DEFAULT_LANGUAGE_CODE));
 
+        return muistutus;
+    }
+
+    private EmailData createEmailData(
+            SahkopostiMuistutus muistutus,
+            HenkiloRestDto henkilo,
+            Oikeustulkki oikeustulkki,
+            LocalDate expiryDate
+    ) throws IOException {
+        EmailMessage email = createEmailMessage(oikeustulkki, expiryDate, muistutus.getLahettaja());
+        EmailRecipient recipient = createExpiryNotificationRecipient(muistutus, henkilo);
+
         EmailData emailData = new EmailData();
+        emailData.setEmail(email);
+        emailData.getRecipient().add(recipient);
+
+        return emailData;
+    }
+
+    private EmailMessage createEmailMessage(
+            Oikeustulkki oikeustulkki,
+            LocalDate expiryDate,
+            String sender
+    ) throws IOException {
         EmailMessage email = new EmailMessage();
+
+        String subject = "Rekisteröintisi on päättymässä";
+        String body = getEmailBody(oikeustulkki, expiryDate, sender);
+
+        email.setSubject(subject);
+        email.setBody(body);
         email.setCallingProcess(callingProcess);
-        email.setFrom(senderEmail);
-        email.setSender(senderEmail);
+        email.setFrom(sender);
+        email.setSender(sender);
         email.setReplyTo(replyTo);
-        email.setTemplateName(templateName);
         email.setLanguageCode(DEFAULT_LANGUAGE_CODE);
         email.setCharset("UTF-8");
         email.setHtml(true);
-        emailData.setEmail(email);
+
+        return email;
+    }
+
+    private String getEmailBody(Oikeustulkki oikeustulkki, LocalDate expiryDate, String sender) throws IOException {
+        String templateName = expiryNotificationTemplateName +  "." + DEFAULT_LANGUAGE_CODE + ".html";
+
+        Map<String, String> params = new TreeMap<>();
+        params.put("expiryDate", firstLanguagePairExpiryDate(oikeustulkki, expiryDate).format(DATE_FORMATTER));
+        params.put("sender", sender);
+
+        EmailTemplateRenderer renderer = new EmailTemplateRendererImpl();
```

Comment:
EmailTemplateRendererImpl on @Service

Reply:
Tiedätkö vastaako Springin @service enemmän ja vähemmän Scalan @Singletonia? Tällöin

```
    @Autowired
    private EmailTemplateRenderer emailTemplateRenderer;
```
varmaan riittäisi yllä, jos täällä palauttaisi
```
return emailTemplateRenderer.renderTemplate(templateName, params);
```

Reply:
En scalasta osaa sanoa, mutta käytännössä se @Service on singleton. Tuollainen muutos riittää.

Reply:
Jes. Ja tietysti Play frameworkistä oli itse asiassa puhe eikä niinkään Scalasta..